pkg/apis/core/validation/validation.go

```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(podAffinityTerm.NamespaceSelector, labelSelectorValidationOptions, fldPath.Child("namespaceSelector"))...)
 	return allErrs
 }
+
+func validateNodeSelectorMutation(fldPath *field.Path, oldNodeSelector, newNodeSelector map[string]string) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node selector was nil, anything can be set.
+	if oldNodeSelector == nil {
+		return allErrs
+	}
+	// Validate no existing node selectors were deleted or mutated.
+	for k, v1 := range oldNodeSelector {
+		if v2, ok := newNodeSelector[k]; !ok || v1 != v2 {
+			allErrs = append(allErrs, field.Invalid(fldPath, newNodeSelector, "only additions to spec.nodeSelector are allowed (no mutations or deletions)"))
+			return allErrs
+		}
+	}
+	return allErrs
+}
+
+func validateNodeAffinityMutation(nodeAffinityPath *field.Path, oldNodeAffinity, newNodeAffinity *core.NodeAffinity) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node affinity was nil, anything can be set.
+	if oldNodeAffinity == nil {
+		return allErrs
+	}
+
+	// RequiredDuringSchedulingIgnoredDuringExecution must not be nil for both old and new nodeAffinity
+	// in order to perform this validation.
+	if oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil || newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return allErrs
+	}
+
+	// NodeSelectorTerms are semantically OR-ed, so adding a new one would
+	// broaden the constraint rather than limit it further.
+	if len(oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) != len(newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) {
```

Comment:
> adding a new one would broaden the constraint rather than limit it further

this isn't correct if len(old) == 0 and len(new) > 0, right? I thought that was allowed

Reply:
Correct, fixed this.